Draft - Rewrite of executionModes.md (#696)
smart-data-lake/smart-data-lake: docs/reference/executionModes.md

title: Execution Modes
---

## Execution modes
Execution modes select the data to be processed by a SmartDataLakeBuilder Action. 
By default, there is no filter applied meaning that every Action reads all data passed on by the predecessor Action. 
If an input has no predecessor Action in the DAG, all data is read from the corresponding DataObject.

This default behavior is applied if you don't set an explicit execution mode.
If you want one of the execution modes described below, you have to explicitly set it on the Action:
```
executionMode {
  type = PartitionDiffMode
  attribute1 = ...
}
```

There are 2 major types of execution modes selecting the subset of data based on:

> * __partitions__: the user selects specific partition to process, or lets SDLB select missing partitions automatically 

Author: Actionally selecting partitions by the user on the command line is not an execution mode. They are just applied as a filter on all start-DataObject of the DAG. This works without setting an execution mode...

```suggestion
* __partition-wise__: Either SDLB will automatically select and process missing partitions from the input, or the partitions are defined manually by a command line parameter.
```

This is in the form of columnName=columnValue, e.g. `--partition-values day=2020-01-01`.
The partition values specified are passed to **all** start inputs (inputs with no predecessor in the DAG). 
In other words, each input DataObject is filtered by the given partition values on the specified column(s).

If the parameters `--partition-values` or `--multi-partition-values` are not specified, SDLB will simply process all available data.

Subsequent Actions will automatically use the same partition values as their predecessor Action.
TODO: what happens to output Data Objects?

:::info
Actions can also have multiple input DataObjects (e.g. CustomDataFrameAction).
In this case, a *main input* needs to be defined to take the partition values from. 
You can either set the main input yourself by specifying the `mainInputId` property on the Action, or SDLB will automatically choose the input.  
Automatic selection uses a heuristic:  
The first input which is not skipped, is not in the list of `inputIdsToIgnoreFilter`, and has the most partition columns defined is chosen.
:::

TODO not clear to me

> On execution every Action takes the partition values of the main input and filters them again for every output DataObject by its defined partition columns, 

Author: This is the case where an output DataObject has different partition columns than the mainInput, e.g. output partition columns are [run_id], but mainInput partition columns are [run_id, mandant_id]. SDLB will remove mandant_id from partition values for writing to output, and pass the reduced partition values to the next Action.

Reviewer: See https://github.com/smart-data-lake/smart-data-lake/pull/696/commits/3725ef1e61b63cc94912e606a607aca34b418b38

which serve again as partition values for the input of the next Action.

This default behaviour is active without providing any explicit `executionMode` in the config of your Actions. 
This means that processing partitioned data is available out-of-the-box in SDLB.

:::info
The partition values act as filter. 

> If you define partition values that do not exist in the input dataObject, these partitions will not be automatically created for you. 

Author: SDLB also validates that partition values exist in input data objects (can be overriden by DataObject.expectedPartitionsCondition). This is especially interesting with CustomDataFrameAction and multiple partitioned inputs, so you dont just join them even if some input data is missing...

Reviewer: See https://github.com/smart-data-lake/smart-data-lake/pull/696/commits/3725ef1e61b63cc94912e606a607aca34b418b38

In the condition, the following attributes are available amongst others to make the decision: `inputPartitionValues`, `outputPartitionValues` and `selectedPartitionValues`.
Use these to fail the run based on expected partitions or time conditions.

Default is `false` meaning that the application of the PartitionDiffMode does not fail the action. 
If there is no data to process, the following actions are skipped.

Example - fail if partitions are not processed in strictly increasing order of partition column `dt`:
```
  failCondition = "(size(selectedPartitionValues) > 0 and array_min(transform(selectedPartitionValues, x -> x.dt)) < array_max(transform(outputPartitionValues, x > x.dt)))"
```

Sometimes the `failCondition` can become quite complex with multiple terms concatenated by or-logic.
To improve interpretability of error messages, multiple fail conditions can be configured as an array using the attribute `failConditions`. 
For each condition you can also define a description which will be inserted into the error message.

<!--TODO show a short example of PartitionDiffMode.failConditions with descriptions -->

#### selectExpression

> The option `selectExpression` defines the list of selected output partitions.

Author: ```suggestion
The option `selectExpression` refines or overrides the list of selected output partitions. It can access the partition values selected by the default behaviour and refine the list, or it can override selected partition values by using input & output partition values directly.
```

The .json file used to store the overall state of your application will be extended with state information for this DataObject.

`DataObjectStateIncrementalMode` needs a main input DataObject that can create incremental output, e.g. `JdbcTableDataObject`, `SparkFileDataObject` or `AirbyteDataObject`. 
The inner meaning of the state is defined by the DataObject and not interpreted outside of it. 
Some DataObjects have a configuration option to define the incremental output, e.g. `JdbcTableDataObject.incrementalOutputExpr`, others just use technical timestamps existing by design, 
e.g. `SparkFileDataObject`.

:::info
This option also comes in handy if you can't use `DataFrameIncrementalMode` because you can't access the output DataObject during initialization.
For example, if you push incremental Parquet files to a remote storage and these files are immediately processed and removed, 
you will find an empty directory and therefore can't consult already uploaded data. 
In this case, SDLB needs to remember what data increments were already uploaded.
:::

### SparkStreamingMode
SDLB also supports a streaming with the `SparkStreamingMode`. 

> Under the hood it uses *Spark Structured Streaming*.

Author: ```suggestion
Under the hood it uses [Spark Structured Streaming](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html).
```

For this mode to work, you need to start SDLB with a `--state-path`, see [Command Line](commandLine.md).
The .json file used to store the overall state of your application will be extended with state information for this DataObject.

`DataObjectStateIncrementalMode` needs a main input DataObject that can create incremental output, e.g. `JdbcTableDataObject`, `SparkFileDataObject` or `AirbyteDataObject`.
The inner meaning of the state is defined by the DataObject and not interpreted outside of it.
Some DataObjects have a configuration option to define the incremental output, e.g. `JdbcTableDataObject.incrementalOutputExpr`, others just use technical timestamps existing by design,
e.g. `SparkFileDataObject`.

:::info
This option also comes in handy if you can't use `DataFrameIncrementalMode` because you can't access the output DataObject during initialization.
For example, if you push incremental Parquet files to a remote storage and these files are immediately processed and removed,
you will find an empty directory and therefore can't consult already uploaded data.
In this case, SDLB needs to remember what data increments were already uploaded.
:::

### SparkStreamingMode

> SDLB also supports a streaming with the `SparkStreamingMode`.

Author: ```suggestion
SDLB also supports streaming with the `SparkStreamingMode`.
```

:::

Spark Structured Streaming is keeping state information about processed data.
To do so, it needs a `checkpointLocation` configured on the SparkStreamingMode.

Note that *Spark Structured Streaming* needs an input DataObject that supports the creation of streaming DataFrames.
For the time being, only the input sources delivered with Spark Streaming are supported.
These are `KafkaTopicDataObject` and all `SparkFileDataObjects`, see also [Spark StructuredStreaming](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#creating-streaming-dataframes-and-streaming-datasets).

### KafkaStateIncrementalMode
A special incremental execution mode for Kafka inputs, remembering the state from the last increment through the Kafka Consumer, 
e.g. committed offsets.

### FileIncrementalMoveMode
Another paradigm for incremental processing with files is to move or delete input files once they are processed.
This can be achieved by using FileIncrementalMoveMode. If option `archiveSubdirectory` is configured, files are moved into that directory after processing, otherwise they are deleted.

> FileIncrementalMoveMode is the only execution mode that can be used with the file engine (see also [Execution engines](executionEngines.md)), but also with SparkFileDataObjects and the data frame engine.

Author: I dont think it "is the only" one, PartitionDiffMode can also be used with file and data frame engine.
```suggestion
FileIncrementalMoveMode can be used with the file engine (see also [Execution engines](executionEngines.md)), but also with SparkFileDataObjects and the data frame engine.
```

Author: Two small comments, otherwise ok to merge :-)
